cgroupsgateway: add support for cpu.shares

thenor opened the pull request:
Add support in CGroupsGateway for configuring cpu.shares.
Add unit tests to verify correct behavior also during whitelisting.

Signed-off-by: Therese Nordqvist <therese.nordqvist@pelagicore.com>

Commit:
cgroupsgateway: Add support for cpu.shares

Add support in CGroupsGateway for configuring cpu.shares.
Add tests to verify correct behavior also during whitelisting
and setting of config.

Signed-off-by: Therese Nordqvist <therese.nordqvist@pelagicore.com>

kursatkobya commented on libsoftwarecontainer/src/gateway/cgroups/cgroupsgateway.h:
```diff
@@ -112,11 +112,24 @@ class JSonError : public CgroupsGatewayError
 };
 
+class InvalidValueError : public CgroupsGatewayError
+{
```
maybe we can combine InvalidValueError & HexFormatError as a invalidValueInput

thenor replied:
Yes, I was thinking of this also. Couldn't decide. I guess the error text describes in more detail anyway.

thenor replied:
@tobsan @breakreturn Perhaps one of you has an opinion about this. It doesn't really make sense, as @kursatkobya points out, to have two different Errors, since they both indicate the same type of error (invalid input).

thenor replied:
@tobsan answered offline that he thinks should combine, so I will do that.

kursatkobya commented on libsoftwarecontainer/src/gateway/cgroups/cgroupsparser.cpp:
```diff
@@ -105,6 +105,36 @@ void CGroupsParser::parseCGroupsGatewayConfiguration(const json_t *element)
         }
+
+    } else if (("cpu.shares" == settingKey)) {
+        long int newValue;
```
probably we don't need to use long here
an integer would be sufficient

thenor replied:
You may be right. There doesn't seem to be a limit, but I didn't know if we wanted to limit the user, but I do agree that an int should be sufficient.

thenor replied:
Ok. I will use int instead. :)

kursatkobya commented on libsoftwarecontainer/src/gateway/cgroups/cgroupsparser.cpp:
```diff
@@ -105,6 +105,36 @@ void CGroupsParser::parseCGroupsGatewayConfiguration(const json_t *element)
+    } else if (("cpu.shares" == settingKey)) {
+        long int newValue;
+        try {
+            newValue = std::stol(settingValue);
```
and so stoi can be used instead

kursatkobya commented on libsoftwarecontainer/src/gateway/cgroups/cgroupsparser.cpp:
```diff
@@ -105,6 +105,36 @@ void CGroupsParser::parseCGroupsGatewayConfiguration(const json_t *element)
+            }
+        }
+        settingValue = std::to_string(newValue);
+        log_debug() << "Value for cpu.shares: " << settingValue;
```
instead of using variable just for debug, it would be better to convert directly  like 
`log_debug() << "Value for cpu.shares: " << std::to_string(newValue);`

thenor replied:
I also need the settingValue for setting the variable further down. So it's not just for the printout.

kursatkobya reviewed (commented):
Look good so far.
Need to be reviewed again after servicetests.
Actually i don't think cgroups needs any component tests. Maybe component test should be removed.

tobsan commented on libsoftwarecontainer/src/gateway/cgroups/cgroupsparser.cpp:
```diff
@@ -105,6 +105,36 @@ void CGroupsParser::parseCGroupsGatewayConfiguration(const json_t *element)
+            throw InvalidValueError(errorMessage);
+        }
+
+        if (newValue < 2) {
```
Why does it have to be at least 2? The shares are relative to each other so any integer value should be fine, no?

thenor replied:
According to https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Resource_Management_Guide/sec-cpu.html : "The value specified in the cpu.shares file must be 2 or higher. "

tobsan commented on libsoftwarecontainer/src/gateway/cgroups/cgroupsparser.cpp:
```diff
@@ -105,6 +105,36 @@ void CGroupsParser::parseCGroupsGatewayConfiguration(const json_t *element)
+        }
+        // if the new value is smaller/equal than the old value, then we don't save the new value.
+        if (m_settings.count(settingKey) != 0) {
+            if (std::stol(m_settings[settingKey]) >= newValue) {
```
If choosing integer, then I guess `std::stoi` should be used here later on as well. Btw, I think the check should be the other way: if the `newValue` is larger, we save it. Giving a smaller value isn't even worthy of a debug output imho :)

thenor replied:
I can remove the debug printout, but not sure how you mean to change it to check that newValue is higher than the one set. I need a return for the case when it is <= newValue, since that implies we won't reach the setting on the variable.

thenor commented:
I talked to @breakreturn about perhaps removing the component test and just keeping service tests and unit tests. So unless @tobsan or other reviewers have strong opinions against that, then I can remove the component test suite for cgroups.

tobsan reviewed (changes requested):
Just small things. If we have proper unit tests and service tests we probably don't need component tests.

Commit:
cgroups: Remove component tests

Removes cgroup component tests since they overlap and don't add any
value compared to service tests.

Signed-off-by: Therese Nordqvist <therese.nordqvist@pelagicore.com>

kursatkobya commented on servicetest/cgroups/test_cgroups.py:
```diff
@@ -214,3 +235,40 @@ def test_netcls_cgroup_set(self):
             sc.terminate()
+
+    def test_cpu_shares_cgroup_set(self):
+        """ TODO: add text
```
more comment would be nice

thenor replied:
Oh thanks for reminding me. I forgot. Good thing I wrote "TODO".

kursatkobya commented on servicetest/cgroups/test_cgroups.py:
```diff
@@ -214,3 +235,40 @@ def test_netcls_cgroup_set(self):
+            sc.terminate()
+
+    def test_cpu_shares_cgroup_whitelisting(self):
+        """ TODO: add text
```
more comment would be nice
